### ISSUE TITLE
Weird interactions between label and milestones editor

### ISSUE THREAD

**Issue by miketaylr** [webcompat/webcompat.com#1820]
Just mess around with setting a milestone or a label... then refresh. Something is weird. Like, setting a milestone will set labels, or remove them? O_o

**Comment by miketaylr** [comment]
Ah, I think this is because if you hit the issue edit route, you just pass in milestone and status... without labels. So it removes labels. We could pass in all labels as well, when editing a milestone? Seems weird, but I don't think there's another way.

GitHub offers a labels API, which allows you to edit labels for an issue, but doesn't offer a dedicated API to modify milestones. You have to use the issue edit API.

@karlcow, thoughts?

**Comment by karlcow** [comment]
I will check the milestone branch today.

**Comment by karlcow** [comment]
ok let's see… I set up the issue:

http://localhost:5000/issues/1277
![capture d ecran 2017-09-13 a 09 19 28](https://user-images.githubusercontent.com/505230/30353983-ad5d218e-9864-11e7-8af9-4eb43d2ac1d2.png)

Requests are:

```
127.0.0.1 - - [13/Sep/2017 09:18:56] "GET /issues/1277 HTTP/1.1" 304 -
127.0.0.1 - - [13/Sep/2017 09:18:57] "GET /api/issues/labels?per_page=100 HTTP/1.1" 304 -
127.0.0.1 - - [13/Sep/2017 09:18:58] "GET /api/issues/1277 HTTP/1.1" 200 -
127.0.0.1 - - [13/Sep/2017 09:19:02] "POST /csp-report HTTP/1.1" 204 -
```

Let me add firefox as a label.

![capture d ecran 2017-09-13 a 09 20 51](https://user-images.githubusercontent.com/505230/30354016-f2734500-9864-11e7-83f0-8fe02dd72855.png)

```
127.0.0.1 - - [13/Sep/2017 09:20:42] "POST /api/issues/1277/labels HTTP/1.1" 200 -

```

* **Note 1:** An additional needstriage has been added. Though on Github we get the right label and milestone.

![capture d ecran 2017-09-13 a 09 24 33](https://user-images.githubusercontent.com/505230/30354089-6f6fdc76-9865-11e7-92b0-9e786f1910a6.png)

The markup is

```
<div class="js-Category-list">
  <span class="wc-Label wc-Label--badge js-Label" style="background-color:#d4c5f9">
    firefox
  </span>

  <span class="wc-Label wc-Label--badge js-Label" style="background-color:#ededed">
    needstriage
  </span>

  <span class="wc-Label wc-Label--badge js-Label" style="background-color:#ff9900">
    needstriage
  </span>
</div>
```

Let's change the milestone to needsdiagnosis. 
AH interesting. 

* **Note 2**: As soon as I open the status widget editor (aka milestones, it sent a post to labels). THIS IS A BUG. It should not call anything.

```
127.0.0.1 - - [13/Sep/2017 09:26:44] "POST /api/issues/1277/labels HTTP/1.1" 200 -
```

What it did is to remove the `browser-firefox` label, because I guess the list is empty. 

![capture d ecran 2017-09-13 a 09 28 36](https://user-images.githubusercontent.com/505230/30354192-07bc4f5a-9866-11e7-8352-a2c0405add44.png)

Then closing the milestone/status editor after selection, Another bug:

* **Note 3**: Closing the status widget editor, call again labels. It should not! This is a bug.

```
127.0.0.1 - - [13/Sep/2017 09:29:40] "POST /api/issues/1277/labels HTTP/1.1" 200 -
127.0.0.1 - - [13/Sep/2017 09:29:42] "PATCH /api/issues/1277/edit HTTP/1.1" 200 -
```

So indeed it changes the status milestone BUT Also change the label. It should not do that.

![capture d ecran 2017-09-13 a 09 31 18](https://user-images.githubusercontent.com/505230/30354239-56992d82-9866-11e7-954d-c255db90cb5c.png)

It doesn't auto-refresh the status too in the current page. It's why you get a misunderstanding of what is happening 

Once we refreshed we get:

![capture d ecran 2017-09-13 a 09 32 21](https://user-images.githubusercontent.com/505230/30354271-8e803628-9866-11e7-8598-69b39f795f63.png)

**Comment by karlcow** [comment]
@miketaylr so indeed we have a bug, not with the /edit route, but with the status/milestone editor calling the labeler route on Opening and Closing.

**Comment by karlcow** [comment]
When I open the status widget editor, it is calling `labels.save` in `issue.js` it should not.
https://github.com/webcompat/webcompat.com/blob/63caeeaaa9c807b39d20090b17860403fb314909/webcompat/static/js/lib/models/issue.js#L131-L143

**Comment by karlcow** [comment]
Oh and why this is called twice here:
https://github.com/webcompat/webcompat.com/blob/63caeeaaa9c807b39d20090b17860403fb314909/webcompat/static/js/lib/models/issue.js#L27-L46

Also opening/closing WITHOUT modifying anything is still calling the /labels/ route. :)

**Comment by karlcow** [comment]
doing a bit more digging. When I comment out

https://github.com/webcompat/webcompat.com/blob/63caeeaaa9c807b39d20090b17860403fb314909/webcompat/static/js/lib/models/issue.js#L127-L143

The issue is solved. We can't set the labels anymore obviously, but at least it means that the Status widget editor was trying to call this.

On the other hand another bug surfaced. With this piece of code commented. When I open the LABEL widget editor, I get, there's a call made to the /edit uri  (This should NOT happen.).

```
 TypeError: statusObject is undefined
[Learn More]
milestones.js:91:11
closeEditor
http://localhost:5000/js/lib/milestones.js:91:11
closeEditor
http://localhost:5000/js/lib/milestones.js:26:5
closeCategoryEditor
http://localhost:5000/js/lib/issues.js:365:7
r
http://localhost:5000/js/vendor/lodash.custom.min.js:26:149
dispatch
http://localhost:5000/js/vendor/jquery-1.11.2.min.js:3:8497
add/r.handle
http://localhost:5000/js/vendor/jquery-1.11.2.min.js:3:5235
```

why does it call milestone. It should not.


```html
<!-- label editor -->
<button 
    class="wc-CategoryEditorLauncher 
           r-ResetButton 
           wc-Icon 
           wc-Icon--gear 
           js-CategoryEditorLauncher 
           js-LabelEditorLauncher" 
           aria-hidden="true" 
           style="display: inline-block;">
    <span class="wc-Accessibility">Edit Labels</span>
</button>

<!-- milestone editor -->
<button 
    class="wc-CategoryEditorLauncher 
           r-ResetButton 
           wc-Icon 
           wc-Icon--gear 
           js-CategoryEditorLauncher 
           js-MilestoneEditorLauncher" 
           aria-hidden="true" 
           style="display: inline-block;">
    <span class="wc-Accessibility">Edit Statuses</span>
</button>
```

**Comment by karlcow** [comment]
@miketaylr Merging your pull request #1821 is not solving the issues.
Opening the editor still call the labels route is there is at least one label checked. At least it doesn't try anymore to re-set a status-* label. So that's better.

**Comment by karlcow** [comment]
Ah  so went on a step by step walk. Just after rendering the MilestoneEditor… the code is going to 

* `closeCategoryEditor` in `issues.js`, 
* then go to `closeEditor` into `labels.js` 🏷 
* then to `updateLabels` into `issue.js`
* then to `initialize` into `label-list.js` (and because the labels editor is close, it doesn't get any meaningful data).
* then go back to `labels.save()` from `updateLabels` into `issue.js` AND IT'S when the POST is made.  

So in the stack there's a wrong branching up there when it's calling `closeCategoryEditor`

Probably because opening will close the other one, but without checking if the other one was already opened in the first place?

in milestones.js
```js
  render: function() {
    this.$el.html(this.template(this.model.toJSON()));
    this.resizeEditorHeight();
    _.defer(
      _.bind(function() {
        this.$el.find(".wc-CategoryEditor-search").focus();
      }, this)
    );
    return this;
  },
```


and in `issues.js`

```js
  closeCategoryEditor: function(e) {
    var target = $(e.target);
    // early return if the editor is closed,
    if (
      // If no category editor is visible
      !this.$el.find(".js-CategoryEditor").is(":visible") ||
      // or we've clicked on the button to open it,
      (target[0].nodeName === "BUTTON" &&
        target.hasClass("js-CategoryEditorLauncher")) ||
      // or clicked anywhere inside the label editor
      target.parents(".js-CategoryEditor").length
    ) {
      // Clicking on one launcher will force to close the other one
      if (
        target[0].nodeName === "BUTTON" &&
        target.hasClass("js-LabelEditorLauncher")
      ) {
        this.milestones.closeEditor();
      } else if (
        target[0].nodeName === "BUTTON" &&
        target.hasClass("js-MilestoneEditorLauncher")
      ) {
        this.labels.closeEditor();
      }
      return;
    } else {
      // Click outside, close both editors
      this.labels.closeEditor();
      this.milestones.closeEditor();
    }
  },
```

**Comment by miketaylr** [comment]
> @miketaylr so indeed we have a bug, not with the /edit route, but with the status/milestone editor calling the labeler route on Opening and Closing.

@karlcow ah, I think I ended up with the same conclusion here: https://github.com/webcompat/webcompat.com/issues/1807#issuecomment-328949852

**Comment by miketaylr** [comment]
> @miketaylr Merging your pull request #1821 is not solving the issues.
> Opening the editor still call the labels route is there is at least one label checked. At least it doesn't try anymore to re-set a status-* label. So that's better.

Yeah, that PR isn't intended to fix the bug. I just noticed it there. It will probably get fixed in #1807.

**Comment by miketaylr** [comment]
(or here)

**Comment by karlcow** [comment]
Closed by #1831
